### PR TITLE
Sapphire - Jennifer Tam

### PR DESCRIPTION
**Pull request by JenniferWTam**

### CODE REVIEW

File: adagrams/game.py [hunk 1]
```diff
@@ -1,11 +1,107 @@
...
+    'R' : 6,
+    'S' : 4,
+    'T' : 6,
+    'U' : 4,
+    'V' : 2,
+    'W' : 2,
+    'X' : 1,
+    'Y' : 2,
+    'Z' : 1
+}
```
Comment (audreyandoy) [inline adagrams/game.py:29]: Nice work using the uppercase naming convention to indicate that this is a constant variable ✨ 

Since `LETTER_POOL` is accessed by one function, it would be reasonable to place them inside the functions rather than as a constant. There are tradeoffs, the structures would clutter the function some, but it keeps the data as close as possible to where it's being used, and would mean other functions couldn't access it to accidentally alter the contents.

File: adagrams/game.py [hunk 2]
```diff
@@ -1,11 +1,107 @@
...
 def draw_letters():
-    pass
+    # Create a list of keys from the LETTER_POOL dictionary and set it to the new variable "letters"
+    letters = list(LETTER_POOL.keys())   
+    # Once all the letters are added to the letters list, the list is shuffled
+    random.shuffle(letters)
+    # First 10 elements in the list are returned
+    return letters[:10]
```
Comment (audreyandoy) [inline adagrams/game.py:36]: Looks like you made ✨Hard Mode✨! Using `list` on a dictionary ( `LETTER_POOL`) creates a list of the dictionary's keys. In this case, you're making a list of all 26 letters with an equal probability of being randomly chosen via `random.choice`. So in other words, (from the `README.md`):

> Since there are 12 `E`s but only 1 `Z`, it should be 12 times as likely for the user to draw an `E` as a `Z`.

The tests do not focus on the statistical accuracy of your code. Instead, the tests check for outcomes that should NOT be possible such as having more than 1 Z in our hand.

Hard Mode is a totally valid way of making this project so don't fret about resubmitting or redoing this (if you have time, you definitely can but it's low-priority compared to the other Learn, PSE, and projects you also need to focus on).

File: adagrams/game.py [hunk 3]
```diff
@@ -1,11 +1,107 @@
...
+    # Loop through all the letters in the word, .upper() returns a string where all characters are in upper case
+    for letter in word.upper():
+        # If letter is not in the word
+        if letter not in letterbank_copy:
+            return False
+        else:
+            # Removes the letter from the 'letters' list and continue looping
+            letterbank_copy.remove(letter)
+    # If all the letters in the word are present in the letter_bank list, return True
+    return True
```
Comment (audreyandoy) [inline adagrams/game.py:50]: 👍  Great use of slicing to make a copy of the `letter_bank`! This is a really clean solution.

File: adagrams/game.py [hunk 4]
```diff
@@ -1,11 +1,107 @@
...
+            winning_word_len = word_len
+        # If the current word's score is equal to the highest score, then  . .
+        elif score == highest_score: 
+            # A check is done to see if the "winning_word_len" is not equal to 10 and if the current word's length is either equal to 10 or less than the "winning_word_len"
+            if winning_word_len != 10 and (word_len == 10 or word_len < winning_word_len):
+                # If this is true, then the new winning word and its length are stored in their respective variables
+                winning_word = word
+                winning_word_len = word_len
+    # Winning word and high score in a tuple
+    return (winning_word, highest_score)
```
Comment (audreyandoy) [inline adagrams/game.py]: 👍 Nice work! This is a really clean solution ✨

File: main.py [hunk 5]
```diff
@@ -33,7 +33,7 @@ def wave_2_run_game():
         display_retry_instructions()
         continue_input = input()
         game_continue = continue_input == "y"
-   
+
```
Comment (audreyandoy) [inline main.py:36]: Looks like a phantom space sneaked into your PR 👻 ! We can avoid adding changes we didn't intend to commit by using `git add file_name.py` to clearly state which file we want to track changes for instead of `git add .` which adds changes from all FILES to a commit.

File: adagrams/game.py [hunk 6]
```diff
@@ -1,11 +1,107 @@
...
+    # If all the letters in the word are present in the letter_bank list, return True
+    return True
 
 def score_word(word):
-    pass
+    # Set initial value of score to zero
+    score = 0
+    # Use the .upper() function to make word upper case
+    word = word.upper()
+    for letter in word:
```
Comment (audreyandoy) [inline adagrams/game.py:57]: We can apply `upper` to word in the `for` loop instead of reassigning `word` in the line prior.

```python
    for letter in word.upper():
```

File: adagrams/game.py [hunk 7]
```diff
@@ -1,11 +1,107 @@
...
+        elif letter in "BCMP":
+            score += 3
+        elif letter in "FHVWY":
+            score += 4
+        elif letter == "K":
+            score += 5
+        elif letter in "JX":
+            score += 8
+        elif letter in "QZ":
+            score += 10
```
Comment (audreyandoy) [inline adagrams/game.py:71]: Harcoding the strings for each point looks neat here but it does have the tradeoff of making this conditional less flexible or easier to edit. For example, what if we decide to add Japanese characters to this game? We'd have to edit each conditional, which can be a tedious process that's more susceptible to accidental changes to the game logic. Instead, we can implement logic that refers to a data structure that associates letters with points. When we need to add Japanese characters, we'd make changes to the data structure ONLY and not the core logic that totals up the score. 

In other words, we should separate referential data (or constant data*) from the core logic of our program.

* constant data is data that isn't being modified by the program

**Review by audreyandoy — COMMENTED** [review]
Awesome job Jennifer ✨

Your submission covers all the learning goals and passes all the tests! You have earned a 🟢 grade!

Great work in making small and frequent commits! Frequent commits and detailed commit messages are super when we need to look back at previous changes and for hunting bugs 🐛

In your PR, you'll find comments on ways to refactor and DRY up your code. Feel free to reach out to me in slack if you have any questions about the feedback I've provided.

Keep up the great work Jennifer✨
